### PR TITLE
Improve fb driver for userspace gui programs

### PR DESCRIPTION
**Pull request by xy-plus**

### CODE REVIEW

File: kernel/src/syscall/fs.rs [hunk 1]
```diff
@@ -155,11 +155,12 @@ impl Syscall<'_> {
...
     ) -> SysResult {
-        info!(
-            "select: nfds: {}, read: {:?}, write: {:?}, err: {:?}, timeout: {:?}",
-            nfds, read, write, err, timeout
-        );
-
+        unsafe {
+            info!(
+                "select: nfds: {}, read: {:?}, write: {:?}, err: {:?}, timeout: {:?}",
+                nfds, read, write, err, *timeout
```
Comment (jiegec) [inline kernel/src/syscall/fs.rs]: timeout should be savely read into kernel until it is printed.

File: kernel/src/syscall/fs.rs [hunk 2]
```diff
@@ -180,6 +181,9 @@ impl Syscall<'_> {
 
         let begin_time_ms = crate::trap::uptime_msec();
         Condvar::wait_events(&[&STDIN.pushed, &(*SOCKET_ACTIVITY)], move || {
+            if nfds as u64 == 0 {
```
Comment (jiegec) [inline kernel/src/syscall/fs.rs]: This check can be put earlier?

File: kernel/src/drivers/gpu/fb.rs [hunk 3]
```diff
@@ -244,10 +244,56 @@ impl Framebuffer {
         var_info.xoffset = self.fb_info.xoffset;
         var_info.yoffset = self.fb_info.yoffset;
         var_info.bits_per_pixel = self.fb_info.depth as u32;
+        var_info.blue = vga::fb_bitfield {
```
Comment (jiegec) [inline kernel/src/drivers/gpu/fb.rs]: It depends on ColorDepth?

File: kernel/src/drivers/gpu/fb.rs [hunk 4]
```diff
@@ -244,10 +244,56 @@ impl Framebuffer {
...
+        };
     }
 
     pub fn fill_fix_screeninfo(&self, fix_info: &mut fb_fix_screeninfo) {
-        fix_info.line_length = self.fb_info.xres * self.fb_info.depth as u32 / 8
+        // pub id: [u8; 16],    /* identification string eg "TT Builtin" */
+
+        // fix_info.smem_start = self.fb_info.vaddr as u64;
+
+        fix_info.smem_start = 0xC0000000 as u64;
```
Comment (jiegec) [inline kernel/src/drivers/gpu/fb.rs]: What's this magic number?

File: kernel/src/drivers/gpu/fb.rs [hunk 5]
```diff
@@ -246,32 +246,30 @@ impl Framebuffer {
         var_info.bits_per_pixel = self.fb_info.depth as u32;
         var_info.blue = vga::fb_bitfield {
             offset: 0 as u32,
-            length: 8 as u32,
+            length: self.fb_info.depth as u32 / 4,
```
Comment (jiegec) [inline kernel/src/drivers/gpu/fb.rs]: This is not so simple, your code won't work for Rgb232 and Rgb565.

**Comment by jiegec** [comment]
You don't need to recreate a pull-request when you are trying to fix it.

**Comment by wangrunji0408** [comment]
I confirm that this version can basically run [musl based MiniGUI v4.0](https://github.com/xy-plus/static-libminigui)!
But only graphic output is working, inputs (keyboard & mouse) is going to be done in the future.

19271572581313_ pic
19301572581367_ pic_hd

**Comment by wangrunji0408** [comment]
@jiegec Plz review again and merge!

**Comment by jiegec** [comment]
Tips: you can use `fbset` to see what these variables actually are in read world.

**Comment by wangrunji0408** [comment]
It still don't work on aarch64 and riscv64 :(

Both of them will get page fault when a new thread calls `pthread_setspecific`.
Seems something went wrong on multi-threading.

The framebuffer and ioctl on `/dev/fb0` are working well.

**Pull request merged**